### DISCUSSION TITLE
Blocks again...can you explain what is happening to my block?

### DISCUSSION THREAD

**Discussion by Bumppoman** [yippee-fun/phlex#458] [General]
Phlex 1.2.1:

```ruby
class Foo < Phlex::HTML
  def template(&)
    instance_eval(&)
  end

  def test_method
    text "Hello!"
  end
end

class Bar
  def template
    Foo.new do
      test_method
    end
  end
end
```

Expected result:  "Hello"
Actual result: "no method error `test_method`..."

Why is the block being evaluated in the context of Bar rather than the context of Foo?

If I use the internal method to take my block back, everything works as expected:

```ruby
class Foo < Phlex::HTML
  def template
    instance_eval(&@_content_block)
  end

  def test_method
    text "Hello!"
  end
end

class Bar
  def template
    Foo.new do
      test_method
    end
  end
end
```

Expected result:  "Hello"
Actual result: "Hello"

Any recommendations here?  Thanks!

**Comment by joeldrapper** [comment]
This is probably because Phlex wraps the block when calling `template` so you can yield it with the benefits of `yield_content`. (`yield_content` will treat the return value as text if it doesn’t do anything to the buffer while being yielded. It's how `h1 { "Hello" }` works.)

`Bar` should be updated to call `test_method` on the yielded object, which is the view unless you explicitly yielded something else. I would pick it up as `|f|` and call `f.test_method`.

```ruby
class Bar < Phlex::HTML
  def template
    render Foo.new do |f|
      f.test_method
    end
  end
end
```

`Foo` can just `yield` the block and it will automatically `yield(self)` within a `yield_content`.

```ruby
class Foo < Phlex::HTML
  def template
    yield
  end

  def test_method
    text "Hello!"
  end
end
```

Alternatively, `Foo` could pass the block to a tag which will yield it.

```ruby
def template(&)
  div(&)
end
```
